File: rvsec/rvsec-android/rvsmart/src/main/java/br/unb/cic/rvsmart/llm/PromptBuilder.java
package br.unb.cic.rvsmart.llm;

import br.unb.cic.rvsmart.core.Config.PromptVersion;
import br.unb.cic.rvsmart.core.ScreenItem;

import java.util.ArrayList;
import java.util.List;

/**
 * Builds the messages list for LLM exploration requests.
 *
 * Entry point: {@link #build(PromptVersion, PromptContext)}.
 * Dispatches to V13 or V17 template based on the version parameter.
 * V12/V14/V15/V16 all fall through to the V13 template.
 *
 * Produces a two-message conversation:
 *   1. System message: role description and available action types
 *   2. User message: screenshot image, numbered UI element list, current activity,
 *      navigation hint, and visited activity summary
 *
 * Prompts are kept concise because Qwen3-VL-4B has limited context capacity.
 * The LLM is instructed to respond with a single tool call in JSON format.
 */
public class PromptBuilder {

    /**
     * Build the messages list for an exploration step.
     *
     * @param version prompt template version; V13 and V17 are fully implemented;
     *                V12/V14/V15/V16 fall through to V13
     * @param ctx     all context data needed to assemble the prompt
     * @return list of messages ready to pass to SglangClient.chat()
     */
    public List<SglangClient.Message> build(PromptVersion version, PromptContext ctx) {
        switch (version) {
            case V17:
                return buildV17(ctx);
            // V12, V13, V14, V15, V16 all use the V13 template
            default:
                return buildV13(ctx);
        }
    }

    // -------------------------------------------------------------------------
    // V13 template (dialog-aware, simple element list)
    // -------------------------------------------------------------------------

    private static final String SYSTEM_V13 =
            "You are an Android UI testing agent. Your task is to explore the app by interacting with UI elements.\n" +
            "DIALOG HANDLING: If you see a permission dialog, click Allow/Accept/OK.\n" +
            "  If you see an error or modal dialog, dismiss it before any other action.\n" +
            "PRIORITY: MOP target elements > navigation to new screens > [UNTESTED] elements > [TESTED] elements.\n" +
            "Available actions:\n" +
            "  click(x, y) — tap an element at normalized coordinates [0,1000)\n" +
            "  long_click(x, y) — long press at normalized coordinates\n" +
            "  scroll(x, y, direction) — scroll at position, direction: up/down/left/right\n" +
            "  type_text(text) — type text into the focused input field\n" +
            "  back() — press the system back button\n" +
            "RULE: Do not click the same position twice in a row.\n" +
            "Respond with exactly one action as JSON: {\"name\": \"<action>\", \"arguments\": {<args>}}";

    private List<SglangClient.Message> buildV13(PromptContext ctx) {
        List<SglangClient.Message> messages = new ArrayList<>();

        // System message: role, dialog instructions, action vocabulary
        messages.add(new SglangClient.Message("system", SYSTEM_V13));

        // User message: screenshot + context text
        List<SglangClient.ContentPart> userParts = new ArrayList<>();

        // Screenshot image first so the model grounds its decision visually
        String imageDataUri = "data:image/jpeg;base64," + ctx.getBase64Screenshot();
        userParts.add(SglangClient.ContentPart.imageUrl(imageDataUri));

        StringBuilder context = new StringBuilder();
        context.append("Current activity: ").append(ctx.getCurrentActivity()).append("\n\n");

        context.append("UI elements:\n");
        List<ScreenItem> items = ctx.getUiElements();
        if (items.isEmpty()) {
            context.append("  (no elements)\n");
        } else {
            int index = 1;
            for (ScreenItem item : items) {
                if (!item.isInteractive()) continue;
                context.append("  ").append(index++).append(". ");
                context.append(simpleClassName(item.getClassName()));
                if (item.getText() != null && !item.getText().isEmpty()) {
                    context.append(" \"").append(truncate(item.getText(), 40)).append("\"");
                } else if (item.getContentDescription() != null
                        && !item.getContentDescription().isEmpty()) {
                    context.append(" [").append(truncate(item.getContentDescription(), 40)).append("]");
                }
                if (item.getBounds() != null) {
                    android.graphics.Rect b = item.getBounds();
                    int cx = (b.left + b.right) / 2;
                    int cy = (b.top + b.bottom) / 2;
                    context.append(" @(").append(cx).append(",").append(cy).append(")");
                }
                context.append("\n");
            }
        }

        String hint = ctx.getNavigationHint();
        if (hint != null && !hint.isEmpty()) {
            context.append("\nNavigation hint: ").append(hint).append("\n");
        }

        if (!ctx.getVisitedActivities().isEmpty()) {
            context.append("\nVisited activities (").append(ctx.getVisitedActivities().size()).append("): ");
            context.append(String.join(", ", ctx.getVisitedActivities())).append("\n");
        }

        context.append("\nIteration: ").append(ctx.getIterationNumber());
        context.append("\nChoose ONE action to explore new UI states or trigger monitored operations.");

        userParts.add(SglangClient.ContentPart.text(context.toString()));
        messages.add(new SglangClient.Message("user", userParts));

        return messages;
    }

    // -------------------------------------------------------------------------
    // V17 template (MOP-aware, test-status tags, action history)
    // -------------------------------------------------------------------------

    private static final String SYSTEM_V17 =
            "You are an Android UI automation assistant.\n" +
            "\n" +
            "REASONING STEPS:\n" +
            "1. SCREEN: Identify screen type (dialog, form, list, menu).\n" +
            "2. DIALOG: If blocking dialog present, handle it first.\n" +
            "3. MOP CHECK: If [DM] or [M] elements are shown, prioritize them.\n" +
            "4. NAVIGATION: Check for actions leading to unvisited screens.\n" +
            "5. ELEMENTS: Select [UNTESTED] element if no navigation or MOP target available.\n" +
            "6. ACTION: Call the action with normalized coordinates [0,1000).\n" +
            "\n" +
            "DIALOG HANDLING:\n" +
            "- Permission dialogs: Click \"Allow\", \"Accept\", \"OK\"\n" +
            "- Error dialogs: Dismiss before other actions\n" +
            "- Use back() if no dismiss button visible\n" +
            "\n" +
            "PRIORITY:\n" +
            "- Elements reaching monitored operations ([DM] direct / [M] transitive) > other actions\n" +
            "- Actions leading to NEW screens > same-screen actions\n" +
            "- [UNTESTED] > [TESTED-Nx] > [WELL-TESTED]\n" +
            "\n" +
            "RULES:\n" +
            "- Do not click the same position consecutively\n" +
            "- If last action had no effect, try a different element\n" +
            "- Explore new screens before testing same screen deeply\n" +
            "\n" +
            "AVOID: navigation bar (bottom), status bar (top)\n" +
            "\n" +
            "Available actions:\n" +
            "  click(x, y) — tap at normalized coordinates [0,1000)\n" +
            "  long_click(x, y) — long press\n" +
            "  scroll(x, y, direction) — direction: up/down/left/right\n" +
            "  type_text(text) — type into focused EditText\n" +
            "  back() — press system back\n" +
            "\n" +
            "Respond with exactly one action as JSON: {\"name\": \"<action>\", \"arguments\": {<args>}}";

    private List<SglangClient.Message> buildV17(PromptContext ctx) {
        List<SglangClient.Message> messages = new ArrayList<>();
        messages.add(new SglangClient.Message("system", SYSTEM_V17));

        List<SglangClient.ContentPart> userParts = new ArrayList<>();
        String imageDataUri = "data:image/jpeg;base64," + ctx.getBase64Screenshot();
        userParts.add(SglangClient.ContentPart.imageUrl(imageDataUri));

        StringBuilder sb = new StringBuilder();

        sb.append("Iteration ").append(ctx.getIterationNumber()).append("\n");

        // Recent actions section (last 5)
        if (!ctx.getRecentActions().isEmpty()) {
            sb.append("Recent actions: ");
            boolean first = true;
            for (br.unb.cic.rvsmart.core.Action a : ctx.getRecentActions()) {
                if (!first) sb.append(", ");
                sb.append(a.getType().name().toLowerCase())
                        .append("@(").append(a.getX()).append(",").append(a.getY()).append(")");
                first = false;
            }
            sb.append("\n");
        }

        sb.append("\nELEMENTS:\n");
        List<br.unb.cic.rvsmart.core.ScreenItem> items = ctx.getUiElements();
        if (items.isEmpty()) {
            sb.append("  (no elements)\n");
        } else {
            int index = 1;
            for (br.unb.cic.rvsmart.core.ScreenItem item : items) {
                if (!item.isInteractive()) continue;
                String elemId = safeElementId(item);
                int count = getInteractionCount(ctx, elemId);
                String statusTag = elementStatusTag(count);
                String mopMarker = elementMopMarker(ctx, elemId);

                sb.append("  ").append(index++).append(". ");
                sb.append(simpleClassName(item.getClassName()));
                if (item.getText() != null && !item.getText().isEmpty()) {
                    sb.append(" \"").append(truncate(item.getText(), 40)).append("\"");
                } else if (item.getContentDescription() != null
                        && !item.getContentDescription().isEmpty()) {
                    sb.append(" [").append(truncate(item.getContentDescription(), 40)).append("]");
                }
                if (item.getBounds() != null) {
                    try {
                        android.graphics.Rect b = item.getBounds();
                        int cx = (b.left + b.right) / 2;
                        int cy = (b.top + b.bottom) / 2;
                        sb.append(" @(").append(cx).append(",").append(cy).append(")");
                    } catch (RuntimeException ignored) {}
                }
                sb.append(" ").append(statusTag);
                if (!mopMarker.isEmpty()) sb.append(" ").append(mopMarker);
                sb.append("\n");
            }
        }

        // Screen info line
        int interacted = ctx.getElementInteractionCounts() != null
                ? ctx.getElementInteractionCounts().size() : 0;
        int total = 0;
        for (br.unb.cic.rvsmart.core.ScreenItem si : items) {
            if (si.isInteractive()) total++;
        }
        sb.append("\nSCREEN: ").append(ctx.getCurrentActivity())
                .append(" | ").append(interacted).append("/").append(total)
                .append(" actions tested | iter #").append(ctx.getIterationNumber()).append("\n");

        // MOP navigation hint
        String hint = ctx.getNavigationHint();
        if (hint != null && !hint.isEmpty()) {
            sb.append("\nMOP NAVIGATION:\n").append(hint).append("\n");
        }

        sb.append("\nSelect action. Prioritize elements reaching monitored operations, then navigation to new screens.");

        userParts.add(SglangClient.ContentPart.text(sb.toString()));
        messages.add(new SglangClient.Message("user", userParts));
        return messages;
    }

    /** Compute element ID from a ScreenItem without requiring Android Rect to be non-stub. */
    private String safeElementId(br.unb.cic.rvsmart.core.ScreenItem item) {
        String resId = item.getResourceId();
        if (resId != null && !resId.isEmpty()) return "res:" + resId;
        if (item.getBounds() != null) {
            try {
                android.graphics.Rect b = item.getBounds();
                return "coords:" + ((b.left + b.right) / 2) + "," + ((b.top + b.bottom) / 2);
            } catch (RuntimeException ignored) {}
        }
        return "unknown:" + System.identityHashCode(item);
    }

    private int getInteractionCount(PromptContext ctx, String elemId) {
        if (ctx.getElementInteractionCounts() == null) return 0;
        Integer c = ctx.getElementInteractionCounts().get(elemId);
        return c != null ? c : 0;
    }

    private String elementStatusTag(int count) {
        if (count == 0) return "[UNTESTED]";
        if (count < 5) return "[TESTED-" + count + "x]";
        return "[WELL-TESTED]";
    }

    private String elementMopMarker(PromptContext ctx, String elemId) {
        if (ctx.getDirectMopElements() != null && ctx.getDirectMopElements().contains(elemId))
            return "[DM]";
        if (ctx.getIndirectMopElements() != null && ctx.getIndirectMopElements().contains(elemId))
            return "[M]";
        return "";
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    /** Return the simple (unqualified) class name for display. */
    private String simpleClassName(String fullName) {
        if (fullName == null) return "View";
        int dot = fullName.lastIndexOf('.');
        return dot >= 0 ? fullName.substring(dot + 1) : fullName;
    }

    /** Truncate a string to at most maxLen characters. */
    private String truncate(String s, int maxLen) {
        if (s == null) return "";
        return s.length() <= maxLen ? s : s.substring(0, maxLen - 1) + "…";
    }
}

File: rv-android/modules/aperv-llm-validation/src/aperv_llm_validation/pipeline/prompt_builder.py
"""Prompt construction for APE-RV LLM validation pipeline.

Replicates the prompt building logic from Java's ApePromptBuilder (commit b2852dd).
Produces OpenAI-format messages and tool schemas for the SGLang inference server.
"""

from __future__ import annotations

from aperv_llm_validation.constants import QWEN_COORD_RANGE
from aperv_llm_validation.data.models import PromptConfig, Widget


def _simple_class_name(full_name: str) -> str:
    """Extract simple class name: 'android.widget.Button' -> 'Button'."""
    if not full_name:
        return "View"
    dot = full_name.rfind(".")
    return full_name[dot + 1 :] if dot >= 0 else full_name


def _truncate(s: str, max_len: int) -> str:
    """Truncate string to max_len, appending ellipsis if truncated."""
    if not s:
        return ""
    if len(s) <= max_len:
        return s
    return s[: max_len - 1] + "\u2026"


MAX_TEXT_LEN = 50


def build_widget_list(widgets: list[Widget], device_w: int, device_h: int) -> str:
    """Format widgets as APE-style indexed list.

    Format: [i] ClassName "text" @(normX,normY) (v:0)
    Where normX = int((centerX / device_w) * 1000), clamped to [0, 999].
    v:0 always (offline, no visit history).
    """
    lines: list[str] = []
    for i, w in enumerate(widgets):
        class_simple = _simple_class_name(w.class_name)
        cx, cy = w.center

        norm_x = int((cx / device_w) * QWEN_COORD_RANGE)
        norm_y = int((cy / device_h) * QWEN_COORD_RANGE)
        norm_x = max(0, min(norm_x, QWEN_COORD_RANGE - 1))
        norm_y = max(0, min(norm_y, QWEN_COORD_RANGE - 1))

        # Display text: prefer text, fallback to content_desc
        display_text = w.text if w.text else w.content_desc
        truncated = _truncate(display_text, MAX_TEXT_LEN)

        lines.append(f'[{i}] {class_simple} "{truncated}" @({norm_x},{norm_y}) (v:0)')

    return "\n".join(lines)


def build_system_message(include_type_text: bool, include_reasoning: bool = False) -> str:
    """Build the system message — exact replica of Java's ApePromptBuilder.buildSystemMessage().

    The include_reasoning parameter is accepted for API compatibility but does not
    affect the system message text (reasoning is handled in the tool schema only).
    """
    parts: list[str] = []
    parts.append("You are an Android UI testing agent exploring an app.\n")
    parts.append("DIALOG: If permission/error dialog visible, dismiss it first (click Allow/OK).\n")
    parts.append("PRIORITY: [DM]/[M] elements > unvisited (v:0) > visited.\n")
    parts.append("AVOID: status bar (top), navigation bar (bottom).\n")
    parts.append("RULES: Don't click same position twice. ")
    parts.append("Use type_text for input fields with valid data ")
    parts.append("(email: user@example.com, password: Test1234!, domain: example.com, search: relevant term).\n")
    parts.append("Tools (coordinates in [0,1000) normalized space):\n")
    parts.append("  click(x, y) \u2014 tap element\n")
    parts.append("  long_click(x, y) \u2014 long press element\n")
    if include_type_text:
        parts.append("  type_text(x, y, text) \u2014 type into field\n")
    parts.append("  back() \u2014 press back\n")
    parts.append('Respond with one JSON: {"name": "<action>", "arguments": {<args>}}')
    return "".join(parts)


def build_user_text(
    activity: str,
    widgets: list[Widget],
    device_w: int,
    device_h: int,
) -> str:
    """Build the user text part: screen header + widget list."""
    # Simple activity name
    if not activity:
        activity_simple = "Unknown"
    else:
        dot = activity.rfind(".")
        activity_simple = activity[dot + 1 :] if dot >= 0 else activity

    parts: list[str] = []
    parts.append(f'Screen "{activity_simple}":')
    widget_list = build_widget_list(widgets, device_w, device_h)
    if widget_list:
        parts.append(widget_list)

    return "\n".join(parts)


def build_messages(
    screenshot_b64: str,
    widgets: list[Widget],
    activity: str,
    device_w: int,
    device_h: int,
    prompt_config: PromptConfig | None = None,
) -> list[dict]:
    """Build the 2-message prompt array: [system, user].

    User message has 2 content parts: [image_url, text].
    image_url format: data:image/jpeg;base64,{b64}
    """
    # Determine if input fields present
    has_input = any(w.editable for w in widgets)

    # System message
    system_text = build_system_message(include_type_text=has_input)

    # User text
    user_text = build_user_text(activity, widgets, device_w, device_h)

    # Image URL
    image_url = f"data:image/jpeg;base64,{screenshot_b64 or ''}"

    return [
        {"role": "system", "content": system_text},
        {
            "role": "user",
            "content": [
                {"type": "image_url", "image_url": {"url": image_url}},
                {"type": "text", "text": user_text},
            ],
        },
    ]


def build_tool_schema(include_type_text: bool, include_reasoning: bool = False) -> list[dict]:
    """Build OpenAI tools array with click, long_click, back, and conditionally type_text.

    If include_reasoning is True, each tool gets an optional 'reasoning' string parameter.
    """

    def _coord_params(include_reasoning: bool) -> dict:
        props = {
            "x": {"type": "integer", "description": "Normalized x coordinate [0, 1000)"},
            "y": {"type": "integer", "description": "Normalized y coordinate [0, 1000)"},
        }
        required = ["x", "y"]
        if include_reasoning:
            props["reasoning"] = {"type": "string", "description": "Reasoning for this action"}
        return {"type": "object", "properties": props, "required": required}

    tools = [
        {
            "type": "function",
            "function": {
                "name": "click",
                "description": "Tap element at coordinates",
                "parameters": _coord_params(include_reasoning),
            },
        },
        {
            "type": "function",
            "function": {
                "name": "long_click",
                "description": "Long press element at coordinates",
                "parameters": _coord_params(include_reasoning),
            },
        },
    ]

    if include_type_text:
        type_text_props = {
            "x": {"type": "integer", "description": "Normalized x coordinate [0, 1000)"},
            "y": {"type": "integer", "description": "Normalized y coordinate [0, 1000)"},
            "text": {"type": "string", "description": "Text to type into the field"},
        }
        required = ["x", "y", "text"]
        if include_reasoning:
            type_text_props["reasoning"] = {"type": "string", "description": "Reasoning for this action"}
        tools.append(
            {
                "type": "function",
                "function": {
                    "name": "type_text",
                    "description": "Type text into input field at coordinates",
                    "parameters": {"type": "object", "properties": type_text_props, "required": required},
                },
            }
        )

    back_props: dict = {}
    back_required: list[str] = []
    if include_reasoning:
        back_props["reasoning"] = {"type": "string", "description": "Reasoning for this action"}
    tools.append(
        {
            "type": "function",
            "function": {
                "name": "back",
                "description": "Press back button",
                "parameters": {"type": "object", "properties": back_props, "required": back_required},
            },
        }
    )

    return tools
